x/cdp/keeper/auctions.go

```diff
@@ -0,0 +1,189 @@
+		// start an auction for one lot, attempting to raise depositDebtAmount
+		_, err := k.auctionKeeper.StartCollateralAuction(
+			ctx, types.LiquidatorMacc, sdk.NewCoin(dep.Amount[0].Denom, auctionSize), sdk.NewCoin(principalDenom, depositDebtAmount), []sdk.AccAddress{dep.Depositor},
+			[]sdk.Int{auctionSize}, sdk.NewCoin(k.GetDebtDenom(ctx), depositDebtAmount))
+		if err != nil {
+			panic(err)
+		}
+		// subtract one lot's worth of debt from the total debt covered by this deposit
+		*debt = debt.Sub(depositDebtAmount)
+		// update the deposits collateral amount
+		dep.Amount[0].Amount = dep.Amount[0].Amount.Sub(auctionSize)
+		// update the total collateral
+		*totalCollateral = totalCollateral.Sub(auctionSize)
+	}
+}
+
+// CreateAuctionFromPartialDeposits creates an auction from the input partial deposits
+func (k Keeper) CreateAuctionFromPartialDeposits(ctx sdk.Context, partialDeps partialDeposits, debt *sdk.Int, collateral *sdk.Int, auctionSize sdk.Int, bidDenom string) {
+	returnAddrs := []sdk.AccAddress{}
+	returnWeights := []sdk.Int{}
+	for _, pd := range partialDeps {
+		returnAddrs = append(returnAddrs, pd.Depositor)
+		returnWeights = append(returnWeights, pd.DebtShare)
+	}
+	_, err := k.auctionKeeper.StartCollateralAuction(ctx, types.LiquidatorMacc, sdk.NewCoin(partialDeps[0].Amount[0].Denom, auctionSize), sdk.NewCoin(bidDenom, partialDeps.SumDebt()), returnAddrs, returnWeights, sdk.NewCoin(k.GetDebtDenom(ctx), partialDeps.SumDebt()))
+	if err != nil {
+		panic(err)
+	}
+	*debt = debt.Sub(partialDeps.SumDebt())
+	*collateral = collateral.Sub(partialDeps.SumCollateral())
+}
+
+// NetSurplusAndDebt burns surplus and debt coins equal to the minimum of surplus and debt balances held by the liquidator module account
+// for example, if there is 1000 debt and 100 surplus, 100 surplus and 100 debt are burned, netting to 900 debt
+func (k Keeper) NetSurplusAndDebt(ctx sdk.Context) {
+	totalSurplus := k.GetTotalSurplus(ctx, types.LiquidatorMacc)
+	debt := k.GetTotalDebt(ctx, types.LiquidatorMacc)
+	netAmount := sdk.MinInt(totalSurplus, debt)
+	if netAmount.GT(sdk.ZeroInt()) {
+		surplusToBurn := netAmount
+		err := k.supplyKeeper.BurnCoins(ctx, types.LiquidatorMacc, sdk.NewCoins(sdk.NewCoin(k.GetDebtDenom(ctx), netAmount)))
+		if err != nil {
+			panic(err)
+		}
+		for _, dp := range k.GetParams(ctx).DebtParams {
+			for surplusToBurn.GT(sdk.ZeroInt()) {
+				balance := k.supplyKeeper.GetModuleAccount(ctx, types.LiquidatorMacc).GetCoins().AmountOf(dp.Denom)
+				if balance.LT(netAmount) {
+					err = k.supplyKeeper.BurnCoins(ctx, types.LiquidatorMacc, sdk.NewCoins(sdk.NewCoin(dp.Denom, balance)))
+					if err != nil {
+						panic(err)
+					}
+					surplusToBurn = surplusToBurn.Sub(balance)
+				} else {
+					err = k.supplyKeeper.BurnCoins(ctx, types.LiquidatorMacc, sdk.NewCoins(sdk.NewCoin(dp.Denom, surplusToBurn)))
+					if err != nil {
+						panic(err)
+					}
+					surplusToBurn = sdk.ZeroInt()
+				}
+			}
```

Comment:
It looks like this can get caught in a loop on the first dp (if it's balance is < netAmount)

Reply:
Good catch! The outer and inner loops should be switched.